### PR TITLE
FastSpeech

### PR DESCRIPTION
**Pull request by stasbel**
Complete FastSpeech model with docs, tests and tutorial.

### CODE REVIEW

File: examples/tts/fastspeech.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
+# Copyright (c) 2019 NVIDIA Corporation
```
Comment (okuchaiev) [inline examples/tts/fastspeech.py]: please put appropriate copyright header

File: examples/tts/fastspeech_durations.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
+# Copyright (c) 2019 NVIDIA Corporation
```
Comment (okuchaiev) [inline examples/tts/fastspeech_durations.py]: same, here and (perhaps in other new files which you've added) - use the standard licence/copyright header

File: nemo/collections/tts/fastspeech/data.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        shuffle (bool): See PyTorch DataLoader.
+            Defaults to True.
+        num_workers (int): See PyTorch DataLoader.
+            Defaults to 0.
+        perturb_config (dict): Currently disabled.
+
+    """
+
+    @property
+    def output_ports(self) -> Optional[Dict[str, NeuralType]]:
```
Comment (okuchaiev) [inline nemo/collections/tts/fastspeech/data.py]: when you are defining new neural modules, please put "add_port_docs" decorator like here https://github.com/NVIDIA/NeMo/blob/master/nemo/collections/tts/tacotron2_modules.py#L37 so that nicer autodocs would be generated

Comment (okuchaiev) [reply nemo/collections/tts/fastspeech/data.py]: the same comment is probably true for all your other input/output ports

File: nemo/collections/tts/fastspeech/loss.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
+# Copyright (c) 2019 NVIDIA Corporation
```
Comment (okuchaiev) [inline nemo/collections/tts/fastspeech/loss.py]: same comments: proper header, and add_port_docs decorator for input/output ports

File: nemo/collections/tts/fastspeech/loss.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    """Neural Module Wrapper for Fast Speech Loss.
+
+    Calculates final loss as sum of two: MSE for mel spectrograms and MSE for durations.
+
+    """
+
+    @property
+    def input_ports(self) -> Optional[Dict[str, NeuralType]]:
+        return dict(
+            mel_true=NeuralType(('B', 'D', 'T'), MelSpectrogramType()),
```
Comment (okuchaiev) [inline nemo/collections/tts/fastspeech/loss.py]: a question: does FastSpeech work with MelSpectrogram only or with MFCC as well?

File: nemo/collections/tts/fastspeech/model.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
+import collections
```
Comment (okuchaiev) [inline nemo/collections/tts/fastspeech/model.py]: you missed header here.
Also, could you please move all pytorch classes under tts/parts ?

File: nemo/collections/tts/fastspeech_modules.py [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+        num_workers (int): See PyTorch DataLoader.
+            Defaults to 0.
+        perturb_config (dict): Currently disabled.
+
+    """
+
+    @property
+    @add_port_docs
+    def output_ports(self):
+        """Returns definitions of module output ports."""
```
Comment (blisc) [inline nemo/collections/tts/fastspeech_modules.py]: @okuchaiev Do we need this docstring?

File: nemo/collections/tts/parts/tacotron2.py [hunk 8]
```diff
@@ -488,7 +488,7 @@ def infer(self, memory, memory_lengths):
             alignments += [alignment]
 
             if len(mel_outputs) == self.max_decoder_steps:
-                logging.warning("Reached max decoder steps")
+                nemo.logging.warning(f"Reached max decoder steps {self.max_decoder_steps}")
```
Comment (blisc) [inline nemo/collections/tts/parts/tacotron2.py]: please switch back to logging.warning. Also please don't use format strings in logging messages.

File: nemo/collections/tts/parts/transformer/conv_transpose.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from torch import nn
+
+
+class Conv(nn.Module):
```
Comment (blisc) [inline nemo/collections/tts/parts/transformer/conv_transpose.py]: Do we really need this class? Are there no equivalent classes anywhere else in the repo?

File: nemo/collections/tts/parts/transformer/sublayers.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+# limitations under the License.
+
+import numpy as np
+import torch.nn.functional as F
+from torch import nn
+
+from nemo.collections.tts.parts.transformer import modules
+
+
+class MultiHeadAttention(nn.Module):
```
Comment (blisc) [inline nemo/collections/tts/parts/transformer/sublayers.py]: Ok I know for sure, this class already exists. Please use the nlp module

File: setup.py [hunk 11]
```diff
@@ -28,6 +28,19 @@
...
+    __description__,
+    __download_url__,
+    __homepage__,
+    __keywords__,
+    __license__,
+    __package_name__,
+    __repository_url__,
+    __version__,
+)
+
```
Comment (blisc) [inline setup.py]: I have no idea how you're not breaking CI with this change. But this change will break setup.py.

File: setup.py [hunk 12]
```diff
@@ -128,6 +128,12 @@ class StyleCommand(distutils_cmd.Command):
         ('fix', None, 'True if tries to fix issues in-place.'),
     ]
 
+    def __init__(self, dist):
+        super().__init__(dist)
+
+        # Nasty trick to look to formatters programs in python bin folder.
+        os.environ['PATH'] += os.pathsep + os.path.dirname(sys.executable)
+
```
Comment (blisc) [inline setup.py]: Why is this necessary?

File: tests/tts/test_tts.py [hunk 13]
```diff
@@ -189,3 +192,95 @@ def test_waveglow_training(self):
         optimizer.train(
             [loss_t], callbacks=[callback], optimizer="sgd", optimization_params={"num_epochs": 10, "lr": 0.0003},
         )
+
+    def test_fastspeech(self):
```
Comment (blisc) [inline tests/tts/test_tts.py]: Let's sit on this change until after we merge #430

Comment (blisc) [reply tests/tts/test_tts.py]: Can you add a test label that is compatible with #454

File: examples/tts/fastspeech.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+        log_dir=work_dir / 'log',
+        checkpoint_dir=work_dir / 'checkpoints',
+        tensorboard_dir=work_dir / 'tensorboard',
+        files_to_copy=[args.model_config],
+    )
+
+    yaml_loader = yaml.YAML(typ="safe")
+    with open(args.model_config) as f:
+        config = attrdict.AttrDict(yaml_loader.load(f))
+    nemo.logging.info(f'Config: {config}')
```
[the remark this answers is not in the document]
Comment (stasbel) [reply examples/tts/fastspeech.py]: Why? What if user going to mix up python and nemo logging within one program?

File: nemo/collections/asr/parts/dataset.py [hunk 15]
```diff
@@ -135,9 +135,7 @@ def __init__(
     ):
         self.collection = collections.ASRAudioText(
             manifests_files=manifest_filepath.split(','),
-            parser=parsers.ENCharParser(
-                labels=labels, unk_id=unk_index, blank_id=blank_index, do_normalize=normalize,
-            ),
+            parser=parsers.CharParser(labels=labels, unk_id=unk_index, blank_id=blank_index, do_normalize=normalize,),
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/asr/parts/dataset.py]: To be compatible with TranscriptDataset (this is needed for FastSpeech code to work).

File: nemo/collections/tts/fastspeech_modules.py (hunk 7 above)
Comment (stasbel) [reply nemo/collections/tts/fastspeech_modules.py]: You mean "Returns definitions of module output ports."? Pretty much every NeMo module has it right now.

File: nemo/collections/tts/parts/transformer/conv_transpose.py (hunk 9 above)
Comment (stasbel) [reply nemo/collections/tts/parts/transformer/conv_transpose.py]: Conv class added transposes for input tensor. Seems like this is need for implementing DurationPredictor as sequence of modules.
Not sure about other such things in NeMo.

File: nemo/collections/tts/parts/transformer/sublayers.py (hunk 10 above)
Comment (stasbel) [reply nemo/collections/tts/parts/transformer/sublayers.py]: Well, input, internals and output are slightly different for other class. I am not sure we can merge them w/o problems.

File: setup.py (hunk 11 above)
Comment (stasbel) [reply setup.py]: Fixed.
There is an error in isort: when style scope is just single file, isort doesn't respect the setup.cfg setting for some reason.

File: setup.py (hunk 12 above)
Comment (stasbel) [reply setup.py]: Deleted.

File: tests/tts/test_tts.py (hunk 13 above)
Comment (stasbel) [reply tests/tts/test_tts.py]: Done.

File: examples/tts/fastspeech.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply examples/tts/fastspeech.py]: Tensorflow, for example, uses `tf.logging` as standard.

File: nemo/collections/asr/parts/dataset.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/asr/parts/dataset.py]: Both datasets uses base CharParser now.

**Comment by stasbel** [comment]
FastSpeech transformer seems to be somewhat custom, not sure if merging it with nlp part is needed.

File: examples/tts/fastspeech.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (blisc) [reply examples/tts/fastspeech.py]: If the user wants to do that, they can do that. For things inside our repo, please stick to the nemo logging

File: nemo/collections/asr/parts/dataset.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/asr/parts/dataset.py]: Make it an option and default it to EN

File: nemo/collections/tts/fastspeech_modules.py (hunk 7 above)
Comment (blisc) [reply nemo/collections/tts/fastspeech_modules.py]: Yes but it's a useless docstring so why bother adding it.

File: nemo/collections/tts/parts/transformer/conv_transpose.py (hunk 9 above)
Comment (blisc) [reply nemo/collections/tts/parts/transformer/conv_transpose.py]: Did you remove this file?

File: nemo/collections/tts/parts/transformer/sublayers.py (hunk 10 above)
Comment (blisc) [reply nemo/collections/tts/parts/transformer/sublayers.py]: Please find a way to merge them. I do not see the point in creating multiple transformer modules.

File: tests/tts/test_tts.py (hunk 13 above)
Comment (blisc) [reply tests/tts/test_tts.py]: Unclassified is not an option.
I would mark this as integration or system. Depending on what you want to test. Since you made a unittest, I assume it's more of a integration test, in which case you should only run it for 1 or 2 steps.

If you want to change it to a system test, please drop num_epochs to be between 1-3 and check that loss or other metric acts as expected.

File: nemo/collections/tts/fastspeech_modules.py (hunk 7 above)
Comment (blisc) [reply nemo/collections/tts/fastspeech_modules.py]: This is not blocking your PR, but a general question posed at @okuchaiev

File: nemo/collections/tts/parts/transformer/layers.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+        output = output.view(n_head, sz_b, len_q, d_v)
+        output = output.permute(1, 2, 0, 3).contiguous().view(sz_b, len_q, n_head * d_v)  # b x lq x (n*dv)
+
+        output = self.dropout(self.fc(output))
+        output = self.layer_norm(output + residual)
+
+        return output, attn
+
+
+class PositionwiseFeedForward(nn.Module):
```
Comment (blisc) [inline nemo/collections/tts/parts/transformer/layers.py]: If you truly want to create your class, please tell me what the differences are between this class and this one in nemo_nlp: https://github.com/NVIDIA/NeMo/blob/master/nemo/collections/nlp/nm/trainables/common/transformer/transformer_modules.py#L274

File: examples/tts/fastspeech.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply examples/tts/fastspeech.py]: Done.

File: nemo/collections/asr/parts/dataset.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/asr/parts/dataset.py]: Done.

File: nemo/collections/tts/parts/transformer/models.py [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+        enc_output, enc_slf_attn = self.slf_attn(enc_input, enc_input, enc_input, mask=slf_attn_mask)
+        enc_output *= non_pad_mask
+
+        enc_output = self.pos_ffn(enc_output)
+        enc_output *= non_pad_mask
+
+        return enc_output, enc_slf_attn
+
+
+class Encoder(nn.Module):
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/parts/transformer/models.py]: Done.

File: nemo/collections/tts/parts/transformer/models.py [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+
+        for i, enc_layer in enumerate(self.layer_stack):
+            enc_output, enc_slf_attn = enc_layer(enc_output, non_pad_mask=non_pad_mask, slf_attn_mask=slf_attn_mask)
+            if return_attns:
+                enc_slf_attn_list += [enc_slf_attn]
+
+        return enc_output, non_pad_mask
+
+
+class Decoder(nn.Module):
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/parts/transformer/models.py]: Done.

File: nemo/collections/tts/parts/transformer/sublayers.py (hunk 10 above)
Comment (stasbel) [reply nemo/collections/tts/parts/transformer/sublayers.py]: Done.

File: nemo/collections/tts/parts/transformer/layers.py (hunk 16 above)
Comment (stasbel) [reply nemo/collections/tts/parts/transformer/layers.py]: Done.

**Comment by blisc** [comment]
Did you push any changes? The last commit I see is f3294a9

File: tests/tts/test_tts.py (hunk 13 above)
Comment (stasbel) [reply tests/tts/test_tts.py]: marked as integration
